### PR TITLE
Feature/update psps burnable mintable

### PR DESCRIPTION
**Pull request by gokuseii**

### CODE REVIEW

File: contracts/tests/psp1155.rs [hunk 1]
```diff
@@ -48,6 +48,8 @@ mod psp1155 {
     pub struct PSP1155Struct {
         #[PSP1155StorageField]
         psp1155: PSP1155Data,
+        // fields for hater logic
+        hated_account: AccountId,
```
Comment (xgreenx) [inline contracts/tests/psp1155.rs:52]: Better to use simple logic in tests. For example, you can add a bool `return_err_on_before`. And return an error)

Also, add test for `after_token_transfer`. Apply the same changes to all tests.

File: examples/psp22_extensions/burnable/lib.rs [hunk 2]
```diff
@@ -4,17 +4,36 @@
...
+        string::String
+    };
 
     #[ink(storage)]
     #[derive(Default, PSP22Storage)]
     pub struct MyPSP22 {
         #[PSP22StorageField]
         psp22: PSP22Data,
+        // fields for hater logic
+        hated_account: AccountId,
```
Comment (xgreenx) [inline examples/psp22_extensions/burnable/lib.rs:18]: You don't need to test `before_token_transfer` and `after_token_transfer` in integration tests. So you don't need to modify example and integration test to test that=)

**Review by xgreenx — CHANGES_REQUESTED** [review]
`_mint` and `_burn` are permissioned functions. So I think you can remove logic related to validation that the owner of the token is the caller of the `_burn`/allowance is set.

File: contracts/token/psp1155/psp1155.rs [hunk 3]
```diff
@@ -234,6 +249,8 @@ impl<T: PSP1155Storage + Flush> PSP1155Internal for T {
             return Ok(())
         }
 
+        self._before_token_transfer(Some(&from), None, &ids_amounts)?;
```
Comment (xgreenx) [inline contracts/token/psp1155/psp1155.rs:252]: You can remove the logic above related to allowance. It is permissioned function so it is okay if it will ignore that.

File: contracts/token/psp22/psp22.rs [hunk 4]
```diff
@@ -270,11 +292,18 @@ impl<T: PSP22Storage + Flush> PSP22Internal for T {
         self.get_mut().balances.insert(account, from_balance);
         self.get_mut().supply -= amount;
         self._emit_transfer_event(Some(account), None, amount);
+
+        self._after_token_transfer(Some(&account), None, &amount)?;
         Ok(())
     }
 
     default fn _burn_from(&mut self, account: AccountId, amount: Balance) -> Result<(), PSP22Error> {
```
Comment (xgreenx) [inline contracts/token/psp22/psp22.rs:300]: You can remove that function

File: contracts/token/psp721/psp721.rs [hunk 5]
```diff
@@ -283,14 +310,22 @@ impl<T: PSP721Storage + Flush> PSP721Internal for T {
...
         self._emit_transfer_event(None, Some(to), id);
+
+        self._after_token_transfer(None, Some(&to), &id)?;
         Ok(())
     }
 
     default fn _burn_from(&mut self, from: AccountId, id: Id) -> Result<(), PSP721Error> {
+        self._before_token_transfer(Some(&from), None, &id)?;
+
         self._remove_token(from, &id)?;
```
Comment (xgreenx) [inline contracts/token/psp721/psp721.rs:325]: The same here, let;s simplify the logic, you don't need to check allowance and ownershiping

File: tests/ownable.tests.ts [hunk 6]
```diff
@@ -23,10 +23,10 @@ describe('MY_OWNABLE', () => {
 
     // Arrange - Alice is not the owner hence minting should fail
     await expect(query.owner()).to.have.output(sender.address)
-    await expect(contract.tx.mint([bnArg(0), 1])).to.eventually.be.fulfilled
+    await expect(contract.tx.mint(sender.address, [bnArg(0), 1])).to.eventually.be.fulfilled
 
     // Act & Assert - Alice can mint a token
-    await expect(fromSigner(contract, alice.address).tx.mint(bnArg(0), 100)).to.eventually.be.rejected
+    await expect(fromSigner(contract, alice.address).tx.mint(alice.address, bnArg(0), 100)).to.eventually.be.rejected
```
Comment (xgreenx) [inline tests/ownable.tests.ts:29]: I think you can simplify the logic here now. 
```rust
contract.tx.mint(alice.address, bnArg(0), 100))
```

Comment (xgreenx) [reply tests/ownable.tests.ts:29]: The same in all tests

File: tests/psp1155/extensions/burnable.tests.ts [hunk 7]
```diff
@@ -62,9 +62,9 @@ describe('MY_PSP1155_BURNABLE', () => {
         await expect(query.balanceOfBatch([[alice.address, token1], [alice.address, token2]]))
             .to.have.output([amount1, amount2])
```
Comment (xgreenx) [inline tests/psp1155/extensions/burnable.tests.ts:63]: After updating of `burn` to allow burn without allowance you need to rewrite that test=)

File: tests/psp22/extensions/burnable.tests.ts [hunk 8]
```diff
@@ -20,20 +20,20 @@ describe('MY_PSP22_BURNABLE', () => {
         await tx.transfer(ALICE.address, 10, []);
 
         // Act - Burn Alice's tokens
-        await fromSigner(contract, ALICE.address).tx.burn(10)
+        await fromSigner(contract, ALICE.address).tx.burn(ALICE.address, 10)
```
Comment (xgreenx) [inline tests/psp22/extensions/burnable.tests.ts:23]: The same here, you can simplify it

File: tests/psp721/extensions/burnable.tests.ts [hunk 9]
```diff
@@ -45,10 +45,9 @@ describe('MY_PSP721_BURNABLE', () => {
 
         await expect(query.balanceOf(sender.address)).to.have.output(3)
 
-        await expect(fromSigner(contract, alice.address).tx.burnFrom(sender.address, bnArg(0)))
+        await expect(fromSigner(contract, alice.address).tx.burn(sender.address, bnArg(0)))
```
Comment (xgreenx) [inline tests/psp721/extensions/burnable.tests.ts:48]: The same idea here, you need to refactor) ANd check other tests please too

File: contracts/token/psp1155/psp1155.rs (hunk 3 above)
Comment (gokuseii) [reply contracts/token/psp1155/psp1155.rs:252]: Are you talking about this?
```rs
if from != operator && !self._is_approved_for_all(from, operator) {
    return Err(PSP1155Error::NotAllowed)
}
```

Comment (xgreenx) [reply contracts/token/psp1155/psp1155.rs:252]: Yes=)

File: contracts/tests/psp22_burnable.rs [hunk 10]
```diff
@@ -184,24 +223,20 @@ mod psp22_burnable {
...
+        assert!(psp22.burn(accounts.alice, amount_to_burn).is_ok());
 
         // Expecting Alice's balance decrease
         assert_eq!(psp22.balance_of(accounts.alice), alice_balance - amount_to_burn);
-        assert_eq!(psp22.allowance(accounts.alice, accounts.bob), 0);
     }
 
     #[ink::test]
-    fn burn_from_fails_if_amount_exceeds_allowance() {
+    fn burn_from_without_allowance() {
```
Comment (xgreenx) [inline contracts/tests/psp22_burnable.rs:239]: You can remove that test=)

File: contracts/tests/psp721.rs [hunk 11]
```diff
@@ -129,18 +166,6 @@ mod psp721 {
...
         assert_eq!(nft.owner_of([1; 32]), None);
-        // Create token Id 2.
-        assert!(nft._mint([1; 32]).is_ok());
-        // Alice owns 1 token.
-        assert_eq!(nft.balance_of(accounts.alice), 1);
-        // Token Id 2 is owned by Alice.
-        assert_eq!(nft.owner_of([1; 32]), Some(accounts.alice));
-        change_caller(accounts.bob);
-        // Bob cannot transfer not owned tokens.
-        assert_eq!(
```
Comment (xgreenx) [inline contracts/tests/psp721.rs:140]: You removed useful test=) Return it back, please + split that one test into two tests

File: contracts/tests/psp721.rs [hunk 12]
```diff
@@ -197,18 +222,16 @@ mod psp721 {
     }
 
     #[ink::test]
-    fn approved_for_all_works() {
```
Comment (xgreenx) [inline contracts/tests/psp721.rs:200]: Return it back, please)

File: contracts/tests/psp721.rs [hunk 13]
```diff
@@ -223,16 +246,10 @@ mod psp721 {
...
-        change_caller(accounts.alice);
-        // Remove operator approval for Bob on behalf of Alice.
-        assert!(nft.set_approval_for_all(accounts.bob, false).is_ok());
-        // Bob is not an approved operator for Alice.
-        assert_eq!(nft.is_approved_for_all(accounts.alice, accounts.bob), false);
     }
 
     #[ink::test]
-    fn not_approved_transfer_should_fail() {
+    fn not_approved_transfer_works() {
```
Comment (xgreenx) [inline contracts/tests/psp721.rs:252]: You need to remove checks for allowance ONLY for `mint` and `burn`, not for `transfer`. `transfer` should work like before=)

**Review by xgreenx — CHANGES_REQUESTED** [review]

File: examples/psp22/lib.rs [hunk 14]
```diff
@@ -19,11 +19,11 @@ pub mod my_psp22 {
...
         fn _before_token_transfer(
             &mut self,
-            _from: &AccountId,
-            to: &AccountId,
+            _from: Option<&AccountId>,
+            to: Option<&AccountId>,
             _amount: &Balance,
         ) -> Result<(), PSP22Error> {
-            if to == &self.hated_account {
+            if to.is_some() && to.unwrap() == &self.hated_account {
```
Comment (sventime) [inline examples/psp22/lib.rs:26]: `if to == Some(&self.hated_account)`

try to avoid unwrap()

File: contracts/token/psp34/psp34.rs [hunk 15]
```diff
@@ -264,51 +303,40 @@ impl<T: PSP34Storage + Flush> PSP34Internal for T {
...
-            && Some(caller) != self.get().token_approvals.get(id).cloned()
-            && !self._approved_for_all(owner, caller)
-        {
-            return Err(PSP34Error::NotApproved)
-        }
         self.get_mut().token_approvals.take(id);
         self.get_mut().token_owner.take(id);
 
-        let from_balance = self.get_mut().owned_tokens_count.get_mut(&from).unwrap().clone();
+        let from_balance = self.get_mut().owned_tokens_count.get_mut(&from).cloned().unwrap_or(1);
```
Comment (sventime) [inline contracts/token/psp34/psp34.rs:309]: this line breaks the logic. You're defaulting balance to 1

File: tests/psp1155/extensions/burnable.tests.ts [hunk 16]
```diff
@@ -39,38 +38,16 @@ describe('MY_PSP1155_BURNABLE', () => {
         await expect(query.balanceOfBatch([[alice.address, token1], [alice.address, token2]]))
             .to.have.output([0, amount2])
 
-        await contract.tx.burn([[token1, amount1], [token2, amount2]])
+        await contract.tx.burn(sender.address, [[token1, amount1], [token2, amount2]])
         await fromSigner(contract, alice.address).tx.setApprovalForAll(sender.address, true)
```
Comment (sventime) [inline tests/psp1155/extensions/burnable.tests.ts:42]: You don't need setApprovalForAll as you removed this check in `burn`. Please update it in **ALL** cases.

File: contracts/token/psp34/psp34.rs [hunk 17]
```diff
@@ -206,15 +214,46 @@ impl<T: PSP34Storage + Flush> PSP34Internal for T {
...
         data: Vec<u8>,
     ) -> Result<(), PSP34Error> {
-        self._before_token_transfer(&from, &to, &id)?;
+        let owner = self.get().token_owner.get(&id).cloned();
+
+        if owner.is_none() {
+            return Err(PSP34Error::TokenNotExists)
+        }
+
+        self._before_token_transfer(Some(&from), Some(&to), &id)?;
```
Comment (sventime) [inline contracts/token/psp34/psp34.rs:223]: Why not to move it before all other checks?

File: contracts/token/psp34/psp34.rs [hunk 18]
```diff
@@ -287,28 +312,31 @@ impl<T: PSP34Storage + Flush> PSP34Internal for T {
...
         self._add_token(to, id.clone())?;
-        self._emit_transfer_event(None, Some(to), id);
+        self._emit_transfer_event(None, Some(to), id.clone());
+
+        self._after_token_transfer(None, Some(&to), &id)?;
         Ok(())
     }
 
     default fn _burn_from(&mut self, from: AccountId, id: Id) -> Result<(), PSP34Error> {
+        if self.get_mut().token_owner.get(&id).is_none() {
```
Comment (sventime) [inline contracts/token/psp34/psp34.rs:330]: Let's move this to function `_check_token_exists` and reuse it in `_transfer_token_from`. 

```
 default fn _check_token_exists(&self, id: &Id) -> Result<&AccountId, PSP34Error> {
        self.get().token_owner.get(id).ok_or(PSP34Error::TokenNotExists)
    }
```

File: contracts/tests/psp1155_burnable.rs [hunk 19]
```diff
@@ -41,61 +75,86 @@ mod psp1155_burnable {
...
         let accounts = accounts();
 
         let mut nft = PSP1155Struct::new();
-        assert!(nft.mint(accounts.bob, token_id_1, token_1_amount).is_ok());
 
         assert_eq!(
-            Err(PSP1155Error::NotAllowed),
-            nft.burn_from(accounts.bob, vec![(token_id_1, token_1_amount)])
+            Err(PSP1155Error::InsufficientBalance),
+            nft.burn(accounts.alice, vec![(token_id_1, burn_amount)])
```
Comment (yarikbratashchuk) [inline contracts/tests/psp1155_burnable.rs:99]:
```suggestion
            nft.burn(accounts.alice, vec![(token_id_1, burn_amount)]),
            Err(PSP1155Error::InsufficientBalance)
```

File: contracts/tests/psp1155_burnable.rs [hunk 20]
```diff
@@ -41,61 +75,86 @@ mod psp1155_burnable {
...
-        let burn_amount = 2;
+        let burn_amount = 21;
+        let accounts = accounts();
 
         let mut nft = PSP1155Struct::new();
 
         assert_eq!(
             Err(PSP1155Error::InsufficientBalance),
-            nft.burn(vec![(token_id_1, burn_amount)])
+            nft.burn(accounts.bob, vec![(token_id_1, burn_amount)])
```
Comment (yarikbratashchuk) [inline contracts/tests/psp1155_burnable.rs:113]: Please, change the order of expressions (burn goes first, error goes second)

File: contracts/tests/psp1155_burnable.rs [hunk 21]
```diff
@@ -41,61 +75,86 @@ mod psp1155_burnable {
...
-    fn burn_from_insufficient_balance() {
-        let token_id_1 = [1; 32];
-        let burn_amount = 21;
+    fn before_token_transfer_should_fail_burn() {
         let accounts = accounts();
-
+        let token_id_1 = [1; 32];
+        let token_1_amount = 1;
+        let token_id_2 = [2; 32];
+        let token_2_amount = 1;
```
Comment (yarikbratashchuk) [inline contracts/tests/psp1155_burnable.rs:123]: you can just have `token_amount` as it is the same for both tokens in this case

File: contracts/tests/psp1155_burnable.rs [hunk 22]
```diff
@@ -41,61 +75,86 @@ mod psp1155_burnable {
...
+        );
+    }
 
+    #[ink::test]
+    fn after_token_transfer_should_fail_burn() {
+        let accounts = accounts();
+        let token_id_1 = [1; 32];
+        let token_1_amount = 1;
+        let token_id_2 = [2; 32];
+        let token_2_amount = 1;
```
Comment (yarikbratashchuk) [inline contracts/tests/psp1155_burnable.rs:145]: same as above

File: contracts/token/psp34/psp34.rs [hunk 23]
```diff
@@ -206,15 +214,46 @@ impl<T: PSP34Storage + Flush> PSP34Internal for T {
         id: Id,
         data: Vec<u8>,
     ) -> Result<(), PSP34Error> {
-        self._before_token_transfer(&from, &to, &id)?;
+        let owner = self.get().token_owner.get(&id).cloned();
+
+        if owner.is_none() {
```
Comment (sventime) [inline contracts/token/psp34/psp34.rs:219]: This can be refactored after you add _check_token_exists:
```
        self._before_token_transfer(Some(&from), Some(&to), &id)?;

        let owner = self._check_token_exists(&id)?;

        let caller = Self::env().caller();
        if owner != &caller
            && Some(caller) != self.get().token_approvals.get(&id).cloned()
            && !self._approved_for_all(owner.clone(), caller)
        {
            return Err(PSP34Error::NotApproved)
        }
```

File: contracts/tests/psp1155_mintable.rs [hunk 24]
```diff
@@ -36,10 +73,53 @@ mod psp1155_mintable {
...
+        assert_eq!(nft.balance_of(accounts.alice, token_id_1), 0);
+        assert!(nft.mint(accounts.alice, vec![(token_id_1, token_1_amount)]).is_ok());
+        assert_eq!(nft.balance_of(accounts.alice, token_id_1), token_1_amount);
+        // Turn on error on _after_token_transfer
+        nft.change_state_err_on_after();
+        // Alice gets an error on _after_token_transfer
+        assert_eq!(
+            nft.mint(accounts.alice, vec![(token_id_2, token_2_amount)]),
+            Err(PSP1155Error::Custom(String::from("Error on _after_token_transfer")))
+        );
```
Comment (yarikbratashchuk) [inline contracts/tests/psp1155_mintable.rs:123]: You don't really need to create token 2 for these tests, do you?

**Review by yarikbratashchuk — CHANGES_REQUESTED** [review]

File: contracts/tests/psp22.rs [hunk 25]
```diff
@@ -268,4 +304,44 @@ mod psp22 {
...
+    #[ink::test]
+    fn before_token_transfer_should_fail_transfer() {
+        // Constructor works.
+        let mut psp22 = PSP22Struct::new(100);
+        let accounts = ink_env::test::default_accounts::<ink_env::DefaultEnvironment>().expect("Cannot get accounts");
+        // Alice can transfer 10 tokens to Bob
+        assert!(psp22
+            .transfer(accounts.bob, 10, Vec::<u8>::new())
+            .is_ok()
+        );
```
Comment (yarikbratashchuk) [inline contracts/tests/psp22.rs:317]: Please run `cargo fmt` after you're done with changes

**Comment by xgreenx** [comment]
@yarikbratashchuk  Please,  put an update of the version to separate pull request

File: contracts/tests/psp1155_burnable.rs [hunk 26]
```diff
@@ -41,61 +75,78 @@ mod psp1155_burnable {
...
-            Err(PSP1155Error::NotAllowed),
-            nft.burn_from(accounts.bob, vec![(token_id_1, token_1_amount)])
+            nft.burn(accounts.alice, vec![(token_id_1, burn_amount)]),
+            Err(PSP1155Error::InsufficientBalance),
         );
     }
 
     #[ink::test]
-    fn burn_insufficient_balance() {
+    fn burn_from_insufficient_balance() {
```
Comment (xgreenx) [inline contracts/tests/psp1155_burnable.rs:104]: I think that test can be removed because balance is insufficient ib both cases(`burn_insufficient_balance` and `burn_from_insufficient_balance`)

File: contracts/tests/psp22_burnable.rs [hunk 27]
```diff
@@ -112,20 +148,25 @@ mod psp22_burnable {
     fn should_not_burn_if_burn_amount_greater_than_account_balance() {
         let initial_balance = 10;
         let mut psp22 = PSP22Struct::new(initial_balance);
+        let accounts = ink_env::test::default_accounts::<ink_env::DefaultEnvironment>().expect("Cannot get accounts");
```
Comment (xgreenx) [inline contracts/tests/psp22_burnable.rs:151]: You can use `accounts()` from `test_utils`

Comment (xgreenx) [reply contracts/tests/psp22_burnable.rs:151]: The same in other tests and files with tests=)

File: contracts/token/psp1155/psp1155.rs [hunk 28]
```diff
@@ -73,12 +73,15 @@ impl<T: PSP1155Storage + Flush> PSP1155 for T {
         self._transfer_guard(operator, from, to)?;
 
         let ids_amounts = vec![(id, amount)];
-        self._before_token_transfer(&from, &to, &ids_amounts)?;
-        self._do_safe_transfer_check(operator, from, to, ids_amounts, data)?;
 
-        self._transfer_from(from, to, id, amount)?;
+        self._do_safe_transfer_check(operator, from, to, ids_amounts.clone(), data)?;
+
+        self._before_token_transfer(Some(&from), Some(&to), &ids_amounts)?;
```
Comment (xgreenx) [inline contracts/token/psp1155/psp1155.rs:79]: Why did we change the ordering of `_before_token_transfer` and `_do_safe_transfer_check`? I think local check should be before the cross-contract call.

File: contracts/token/psp22/psp22.rs [hunk 29]
```diff
@@ -216,11 +228,14 @@ impl<T: PSP22Storage + Flush> PSP22Internal for T {
 
         self._do_safe_transfer_check(from, to, amount, data)?;
 
+        self._before_token_transfer(Some(&from), Some(&to), &amount)?;
```
Comment (xgreenx) [inline contracts/token/psp22/psp22.rs:231]: The same here, why did you change the ordering?

File: tests/psp22/extensions/burnable.tests.ts [hunk 30]
```diff
@@ -13,101 +13,75 @@ describe('MY_PSP22_BURNABLE', () => {
...
         // Act - burn tokens from Alice and Bob
-        await fromSigner(contract, defaultSigner.address).tx.burnFromMany([[ALICE.address, 10], [BOB.address, 10]])
+        await contract.tx.burnFromMany([[alice.address, 10], [bob.address, 10]])
 
         // Assert - ensure needed amount was burnt
-        await expect(query.balanceOf(ALICE.address)).to.have.output(0);
-        await expect(query.balanceOf(BOB.address)).to.have.output(0);
-    })
-
-    it('Fails if do not have an allowance to burn from one of the account', async () => {
```
Comment (xgreenx) [inline tests/psp22/extensions/burnable.tests.ts:76]: You can create an opposite test to check that we can burn without allowance=)

File: contracts/token/psp1155/psp1155.rs (hunk 28 above)
Comment (gokuseii) [reply contracts/token/psp1155/psp1155.rs:79]: @sventime advised to do it as in OpenZeppelin

**Review by yarikbratashchuk — APPROVED** [review]

File: contracts/token/psp22/psp22.rs [hunk 31]
```diff
@@ -266,23 +283,14 @@ impl<T: PSP22Storage + Flush> PSP22Internal for T {
...
-
-        if current_allowance < amount {
-            return Err(PSP22Error::InsufficientAllowance)
-        }
-
-        let new_amount = current_allowance - amount;
-        self._approve_from_to(account, caller, new_amount)?;
-        self._burn(account, amount)
+        self._after_token_transfer(Some(&account), None, &amount)?;
+        Ok(())
```
Comment (xgreenx) [inline contracts/token/psp22/psp22.rs:294]:
```suggestion
        self._after_token_transfer(Some(&account), None, &amount)
```

**Review by xgreenx — APPROVED** [review]
